src/tracing/esp32_diagnostic_trace/DiagnosticEntry.cpp

```diff
@@ -0,0 +1,137 @@
+
+    // Write label
+    if (entry.label != nullptr)
+    {
+        uint32_t labelSize = strlen(entry.label) > kMaxStringValueSize ? kMaxStringValueSize : strlen(entry.label);
+        if (labelSize >= kMaxStringValueSize)
+        {
+            char labelBuffer[kMaxStringValueSize + 1];
+            memcpy(labelBuffer, entry.label, kMaxStringValueSize);
+            labelBuffer[kMaxStringValueSize] = '\0';
+            ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::LABEL), labelBuffer));
+        }
+        else
+        {
+            ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::LABEL), entry.label, labelSize));
+        }
+    }
+
+    // Write value based on type
+    switch (entry.type)
+    {
+    case ValueType::kInvalidType:
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    case ValueType::kCharString:
+        if (entry.stringValue != nullptr)
+        {
+            uint32_t valueSize = strlen(entry.stringValue) > kMaxStringValueSize ? kMaxStringValueSize : strlen(entry.stringValue);
+            if (valueSize >= kMaxStringValueSize)
+            {
+                char valueBuffer[kMaxStringValueSize + 1];
+                memcpy(valueBuffer, entry.stringValue, kMaxStringValueSize);
+                valueBuffer[kMaxStringValueSize] = '\0';
+                ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::VALUE), valueBuffer));
+            }
+            else
+            {
+                ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::VALUE), entry.stringValue, valueSize));
+            }
+        }
+        break;
+
+    case ValueType::kUnsignedInteger:
+        ReturnErrorOnFailure(writer.Put(chip::TLV::ContextTag(DiagTag::VALUE), entry.uintValue));
+        break;
+
+    case ValueType::kSignedInteger:
+        ReturnErrorOnFailure(writer.Put(chip::TLV::ContextTag(DiagTag::VALUE), entry.intValue));
+        break;
+    }
+
+    ReturnErrorOnFailure(writer.EndContainer(DiagnosticOuterContainer));
+    ReturnErrorOnFailure(writer.Finalize());
+    ChipLogDetail(DeviceLayer, "Diagnostic Value written to storage successfully. label: %s\n", entry.label);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Decode(TLVReader & reader, DiagnosticEntry & entry)
+{
+    TLVType containerType;
+    ReturnErrorOnFailure(reader.EnterContainer(containerType));
+
+    // Read timestamp
+    ReturnErrorOnFailure(reader.Next(ContextTag(DiagTag::TIMESTAMP)));
+    ReturnErrorOnFailure(reader.Get(entry.timestamp));
+
+    // Read label
+    ReturnErrorOnFailure(reader.Next(ContextTag(DiagTag::LABEL)));
+    uint32_t labelSize = reader.GetLength();
+    if (labelSize > kMaxStringValueSize)
+    {
+        return CHIP_ERROR_BUFFER_TOO_SMALL;
+    }
+    ReturnErrorOnFailure(reader.GetString(entry.label, kMaxStringValueSize + 1));
+
+    // Read value
+    ReturnErrorOnFailure(reader.Next());
+    switch (entry.type)
```

Comment:
Where is entry.type read? How do we know it is correct for the TLV data set?
It seems we should either read it or switch on the TLV type and set it instead.